apps/nextjs/src/app/api/chat/chatHandler.ts

```diff
@@ -0,0 +1,255 @@
+    useRag: options.useRag,
+    temperature: options.temperature,
+    numberOfLessonPlansInRag: options.numberOfLessonPlansInRag,
+    usePersistence: options.usePersistence,
+    useModeration: options.useModeration,
+  });
+}
+
+function handleConnectionAborted(req: NextRequest) {
+  const abortController = new AbortController();
+
+  req.signal.addEventListener("abort", () => {
+    console.log("Client has disconnected");
+    abortController.abort();
+  });
+  return abortController;
+}
+
+async function handleThreatDetectionError(
+  span: Span,
+  e: AilaThreatDetectionError,
+  userId: string,
+  id: string,
+  prisma: PrismaClientWithAccelerate,
+) {
+  const heliconeErrorMessage = await handleHeliconeError(userId, id, e, prisma);
+  reportErrorTelemetry(span, e, "AilaThreatDetectionError", "Threat detected");
+  return streamingJSON(heliconeErrorMessage);
+}
+
+async function handleAilaAuthenticationError(
+  span: Span,
+  e: AilaAuthenticationError,
+) {
+  reportErrorTelemetry(span, e, "AilaAuthenticationError", "Unauthorized");
+  return new Response("Unauthorized", { status: 401 });
+}
+
+async function handleGenericError(span: Span, e: Error) {
+  reportErrorTelemetry(span, e, e.name, e.message);
+  return streamingJSON({
+    type: "error",
+    message: e.message,
+    value: `Sorry, an error occurred: ${e.message}`,
+  });
+}
+
+async function getUserId(
+  config: Config,
+  span: Span,
+  id: string,
+): Promise<string> {
+  if (config.shouldPerformUserLookup) {
+    const userLookup = await config.handleUserLookup(span, id);
+
+    if (!userLookup) {
+      throw new Error("User lookup failed");
+    }
+
+    if ("failureResponse" in userLookup) {
+      if (userLookup.failureResponse) {
+        throw new Error("User lookup failed: failureResponse received");
+      }
+    }
+
+    if ("userId" in userLookup) {
+      return userLookup.userId;
+    }
+
+    throw new Error("User lookup failed: userId not found");
+  }
+  invariant(config.mockUserId, "User ID is required");
+
+  return config.mockUserId;
+}
+
+async function generateChatStream(
+  aila: Aila,
+  abortController: AbortController,
+) {
+  return tracer.startActiveSpan("chat-aila-generate", async (generateSpan) => {
+    invariant(aila, "Aila instance is required");
+    const result = await aila.generate({ abortController });
+    generateSpan.end();
+    return result;
+  });
+}
+
+async function handleChatException(
+  span: Span,
+  e: unknown,
+  userId: string | undefined,
+  chatId: string,
+  prisma: PrismaClientWithAccelerate,
+): Promise<Response> {
+  if (e instanceof AilaAuthenticationError) {
+    return handleAilaAuthenticationError(span, e);
+  }
+
+  if (e instanceof AilaThreatDetectionError && userId) {
+    return handleThreatDetectionError(span, e, userId, chatId, prisma);
+  }
+
+  if (e instanceof Error) {
+    return handleGenericError(span, e);
+  }
+
+  throw e;
+}
+
+export async function handleChatPostRequest(
```

Comment:
The actual chat post request function now is much simpler and errors are handled in the above methods, which also report telemetry

Reply:
This function is now so easy to follow at a high level. A lot easier to understand the steps involved